### ISSUE TITLE
Crash when release build built locally in Linux due to C++ ABI change between GCC 10 and GCC 12

### ISSUE THREAD

**Issue by doe24** [oven-sh/bun#11261]
### How can we reproduce the crash?

Before start, I'm sorry for my English grammar may not correct.
I use Ubuntu 20.04 LTS, and I have build Bun following contributing guide, I have installed all dependencies that the guild said.
But after I build and debug or I build release and run it show this error.
In bun debug: (In this pic, command 'bun nn.js' is use Bun in global install from npm)
![Screenshot from 2024-05-22 15-34-58](https://github.com/oven-sh/bun/assets/155100391/8475fc31-73d3-4a63-a423-c3bbd29a1eaf)

In bun release build: 
![eiei](https://github.com/oven-sh/bun/assets/155100391/c62a2690-480c-4dc8-bedd-aeddc57d26da)

I don't know why it crash, before these day I build and run successfully every time when I able to install LLVM though this way in my Ubuntu machine:
' wget https://apt.llvm.org/llvm.sh -O - | sudo bash -s -- 16 all '

But now it cannot (now = since 20 May 24, the day that I reinstall Ubuntu) it cannot install though above method, I manually install though this:
' https://github.com/llvm/llvm-project/releases?page=3 '

I try LLVM [16.0.2, 16.0.3, 16.0.4] and try to build 16.0.6 and use in my Ubuntu machine but it failed.
I don't know why.

The JS code that I test it just:
" console.log('Hello world') " and code that generate by bun init.

### Stack Trace (bun.report)

Bun v1.1.10-canary ([`9399b70`](<https://github.com/oven-sh/bun/tree/9399b7013815f825a9713cb97698283260689b78>)) on linux x86_64 [AutoCommand]

Segmentation fault at address 0x0000000A

- [`static_dict.c:0`](<https://github.com/oven-sh/bun/blob/9399b7013815f825a9713cb97698283260689b78/src/deps/brotli/enc/static_dict.c#L0>): `BrotliFindAllStaticDictionaryMatchesFor`
- `??`
- [`tasn_dec.c:791`](<https://github.com/oven-sh/bun/blob/9399b7013815f825a9713cb97698283260689b78/src/deps/boringssl/crypto/asn1/tasn_dec.c#L791>): `asn1_d2i_ex_primitive`
- [`tasn_dec.c:508`](<https://github.com/oven-sh/bun/blob/9399b7013815f825a9713cb97698283260689b78/src/deps/boringssl/crypto/asn1/tasn_dec.c#L508>): `asn1_template_ex_d2i`
- [`tasn_dec.c:0`](<https://github.com/oven-sh/bun/blob/9399b7013815f825a9713cb97698283260689b78/src/deps/boringssl/crypto/asn1/tasn_dec.c#L0>): `asn1_d2i_ex_primitive`
- [`codec.c:0`](<https://github.com/oven-sh/bun/blob/9399b7013815f825a9713cb97698283260689b78/codec.c#L0>): `base64_stream_decode_sse41`
- [`renamer.zig:56`](<https://github.com/oven-sh/bun/blob/9399b7013815f825a9713cb97698283260689b78/src/renamer.zig#L56>): `src.js_printer.NewPrinter`
- [`semver.zig:1466`](<https://github.com/oven-sh/bun/blob/9399b7013815f825a9713cb97698283260689b78/src/install/semver.zig#L1466>): `src.install.lockfile.Package.parseDependency__anon_292224`
- [`lockfile.zig:5286`](<https://github.com/oven-sh/bun/blob/9399b7013815f825a9713cb97698283260689b78/src/install/lockfile.zig#L5286>): `src.install.lockfile.Buffers.writeArray__anon_292742`
- [`bundler.zig:944`](<https://github.com/oven-sh/bun/blob/9399b7013815f825a9713cb97698283260689b78/src/bundler.zig#L944>): `src.bundler.Bundler.transform`

Features: jsc, tsconfig

**Comment by doe24** [comment]
@Electroid I clone source code from main branch "https://github.com/oven-sh/bun".

The last commit when I clone is "https://github.com/oven-sh/bun/commit/9399b7013815f825a9713cb97698283260689b78".

**Comment by doe24** [comment]
Now I have clone source code from main branch "https://github.com/oven-sh/bun".

The last commit when I clone is "https://github.com/oven-sh/bun/commit/d06498bc9671fabfc57f7ffbdf6b6fe146b2b969".

And  now I have changed llvm to version 16.0.6

But it still crash.

![Screenshot from 2024-05-23 14-28-44](https://github.com/oven-sh/bun/assets/155100391/c1a0e6f2-ac2b-4f59-b168-cb5a385306e7)

**Comment by Jarred-Sumner** [comment]
@doe24 it’s a C++ ABI issue. The debug build of WebKit is compiled with an older libc than the one on your machine. You’ll unfortunately need to compile Bun’s WebKit and then pass -DWEBKIT_DIR=path-to-WebKitBuild/Release to the cmake

**Comment by doe24** [comment]
@Jarred-Sumner  I following your instruction, but it still got the same error.

I compile Bun’s WebKit using clang-16 and clang++-16 from this repo:
"https://github.com/oven-sh/WebKit"

and I save all bun-web-kit to release folder, but it still not work.

**Comment by doe24** [comment]
I forgot to tell you about BunWebkit, it run normally after I compile it.

![Screenshot from 2024-05-26 15-29-15](https://github.com/oven-sh/bun/assets/155100391/5f2224ae-631e-417d-9316-801d5c5ec981)

**Comment by billywhizz** [comment]
I am having the same issue on Ubuntu 22.04. For me, the debug build is working fine but I get this crash with the same error message when doing a release build and running the resulting binary - the built jsc binary works fine.

I am pretty sure release builds were working for me locally at some point but cannot get them working now.

> @doe24 it’s a C++ ABI issue. The debug build of WebKit is compiled with an older libc than the one on your machine. You’ll unfortunately need to compile Bun’s WebKit and then pass -DWEBKIT_DIR=path-to-WebKitBuild/Release to the cmake

I'm a bit confused by why an older libc would cause this issue but not something I am expert in. Normally when I see issues like this it is because a library has been built with a *newer* libc than the one i have available. :thinking: 

One thing that might be relevant is that I am on a pretty old (eighth generation Core i5) cpu. Afaict, it should have the required cpu extensions but maybe it is something being picked up/enabled by compiler optimizations in the CI build which is likely running on a much more modern CPU?

I'll have a look this week and see if I can get to the bottom of it. Happy to take any advice.

@doe24 - can you let me know if this is currently breaking for you with both the debug and the release builds or just the release build? :pray:

**Comment by billywhizz** [comment]
@doe24 there is a good description of the "dual ABI" issue here: https://gcc.gnu.org/onlinedocs/libstdc++/manual/using_dual_abi.html

to summarise, bun builds the webkit static library in CI against an older version of libc++. this is because this static library is intended to be used to build release versions of bun and support as many linux distros, including older ones, as possible.

but, when you compile the rest of the bun code on ubuntu 22+ you are compiling and doing the final link against a newer, incompatible, version of libc++. afaik, the problem manifests itself when you try to call C++ std::string methods or pass strings between components of the system that were compiled against the different versions of libc++.

it might be possible to mitigate/resolve this issue at some point by wrapping things in C api's but i don't think that is on the roadmap right now.

in short, to build a release version on ubuntu 22+ rn, you will have to build webkit locally as jarred says above. the other alternative i am looking at is using VSCode dev containers against a docker image based on debian which uses the older libc++. if i make any progress with that i will comment on here.

will close this issue for now. if you not satisfied with the answer, feel free to re-open. :pray:
